src/main/scala/firrtl/ir/IR.scala

```diff
@@ -421,8 +421,29 @@ object Block {
 }
 
 case class Block(stmts: Seq[Statement]) extends Statement {
-  def serialize: String = stmts map (_.serialize) mkString "\n"
-  def mapStmt(f: Statement => Statement): Statement = Block(stmts map f)
+  def serialize: String = {
+    val res = stmts.view.map(_.serialize).mkString("\n")
+    if (res.nonEmpty) res else EmptyStmt.serialize
+  }
+  def mapStmt(f: Statement => Statement): Statement = {
+    val res = new scala.collection.mutable.ArrayBuffer[Statement]()
+    var its = stmts.iterator :: Nil
+    while (its.nonEmpty) {
+      val it = its.head
+      if (it.hasNext) {
+        it.next() match {
+          case EmptyStmt => // flatten out
+          case b: Block =>
+            its = b.stmts.iterator :: its
```

Comment:
I'm playing around with my transform checking work and I noticed that this behavior may be unintuitive. 

Specifically, it seems counterintuitive that doing a `map` with an identity function can actually (superficially) modify the circuit, e.g.:

```scala
/* The following: */
Block(Seq(EmptyStmt, 
          Block(DefWire(NoInfo, "foo", UIntType(IntWidth(1)))))).mapStmt(identity)

/* Returns: */
Block(ArrayBuffer(DefWire(NoInfo, "foo", UIntType(IntWidth(1)))))
```

This has the effect of making transforms which don't have meaningful modifications to the circuit, superficially modify the circuit. Namely, I'm seeing a false positive invalidation in one of my tests where I think `ExpandConnects` is adding a `Block` and then `CInferTypes` is "changing the circuit" by stripping it, even though `CInferTypes` does no real work. Clearly, the issue is `ExpandConnects` adding a block (and my test treating this as a false positive), but the behavior of `mapStmt(identity)` still smells weird to me.

Reply:
> Specifically, it seems counterintuitive that doing a map with an identity function can actually (superficially) modify the circuit

Oh yeah this is totally wrong. Thinking about `Statement` as a `Functor`, this is a violation of the first functor law: Functors must preserve identity morphisms.

I'd also note that there's a pretty bad inconsistency with `foreach`, which while not a violation of a law to my knowledge, has the potential for confusion and nasty bugs.

How should we resolve this?
* We can revert and reproduce the flattening functionality in a different, more explicit way.
* Alternatively we could define `==` (and `hashCode`) on `Block` to match these semantics

Reply:
Explicit functionality via a new transform, `FlattenStatements`, which (1) removes `EmptyStmt` and (2) flattens/joins `Block`s is the most intuitive solution to me. This would be part of one of the very early forms (`Forms.MinimalHighForm` or `Forms.Resolved`?) and I can deal with adding the appropriate invalidations as they're discovered (I'm pretty sure `ExpandConnects` is adding `Block`s...).

Reply:
I'm of two minds.

When I attempted this kind of thing in the past, I just tactically ran the flattening where necessary. The issue was that it actually slowed things down a little bit (like 5%), whereas this approach of overlaying the flattening with traversal has no overhead/perhaps even a slight win. It could be possible to find a to do both: make the flattening opt-in but also overlay with the next run, but that could also be tricky.

Reply:
Can we just "fix" `ExpandConnects` to not emit that unnecessary block?

Reply:
Yes, we should fix that.

Nonetheless, the behavior here is weird. The old behavior was a `map` the new behavior is something like ` flatMap`, but not quite.

You can play around with this at the following Scastie. You can see the old behavior if you switch to Chisel 3.3.1.

https://scastie.scala-lang.org/vlk2fHeUSXmnXsefl71CrA